File: routers/project_progress.py
"""
补贴项目进度跟踪 API
"""
import json
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session
from typing import Optional

from database import get_db
from models import ProjectProgress, SubsidyType, Village, VillageContact

router = APIRouter(prefix="/api/project-progress", tags=["项目进度"])


@router.get("/{subsidy_type_id}")
def list_progress(
    subsidy_type_id: int,
    village_id: Optional[int] = Query(None),
    db: Session = Depends(get_db),
):
    """获取某项目下所有村的进度记录"""
    q = db.query(ProjectProgress).filter(
        ProjectProgress.subsidy_type_id == subsidy_type_id,
        ProjectProgress.village_name != "待分配",
    )
    if village_id:
        q = q.filter(ProjectProgress.village_id == village_id)
    records = q.order_by(ProjectProgress.village_name).all()
    return [_serialize(r) for r in records]


@router.post("/{subsidy_type_id}")
def upsert_progress(
    subsidy_type_id: int,
    data: dict,
    db: Session = Depends(get_db),
):
    """创建或更新某村的进度记录"""
    village_id = data.get("village_id")
    if not village_id:
        return {"error": "缺少 village_id"}, 400

    record = db.query(ProjectProgress).filter(
        ProjectProgress.subsidy_type_id == subsidy_type_id,
        ProjectProgress.village_id == village_id,
    ).first()

    village = db.query(Village).get(village_id)
    village_name = village.village_name if village else str(village_id)

    if record:
        record.person_name = data.get("person_name", record.person_name)
        record.phone = data.get("phone", record.phone)
        if "stages" in data:
            record.stages = json.dumps(data["stages"], ensure_ascii=False)
        if "note" in data:
            record.note = data.get("note")
    else:
        record = ProjectProgress(
            subsidy_type_id=subsidy_type_id,
            village_id=village_id,
            village_name=village_name,
            person_name=data.get("person_name", ""),
            phone=data.get("phone", ""),
            stages=json.dumps(data.get("stages", []), ensure_ascii=False),
            note=data.get("note", ""),
        )
        db.add(record)

    db.commit()
    return _serialize(record)


@router.post("/{subsidy_type_id}/batch")
def batch_upsert(
    subsidy_type_id: int,
    data: dict,
    db: Session = Depends(get_db),
):
    """批量操作：初始化村列表 / 批量更新阶段状态"""
    action = data.get("action", "init")
    if action == "init":
        # 为所有村创建/更新进度记录，自动同步村负责人信息
        # 优先从 VillageContact 取 is_agri_lead 的联系人，其次用 Village 表的 leader
        villages = db.query(Village).filter(Village.village_name != "待分配").order_by(Village.village_name).all()
        # 加载所有村的负责人（从 VillageContact）
        leads = {}
        for c in db.query(VillageContact).filter(VillageContact.is_agri_lead == True).all():
            leads[c.village_id] = c  # 每个村只保留第一个（后端逻辑保证唯一）
        created, updated = 0, 0
        for v in villages:
            lead = leads.get(v.id)
            person_name = lead.name if lead else (v.leader_name or "")
            phone = lead.phone if lead and lead.phone else (v.leader_phone or "")
            exists = db.query(ProjectProgress).filter(
                ProjectProgress.subsidy_type_id == subsidy_type_id,
                ProjectProgress.village_id == v.id,
            ).first()
            if not exists:
                db.add(ProjectProgress(
                    subsidy_type_id=subsidy_type_id,
                    village_id=v.id,
                    village_name=v.village_name,
                    person_name=person_name,
                    phone=phone,
                    stages="[]",
                ))
                created += 1
            else:
                # 同步负责人信息（从 VillageContact 覆盖）
                if person_name:
                    exists.person_name = person_name
                if phone:
                    exists.phone = phone
                updated += 1
        db.commit()
        return {"ok": True, "created": created, "updated": updated}

    if action == "batch_stage":
        # 批量设置某阶段状态
        stage_name = data.get("stage_name", "")
        status = data.get("status", "done")
        for row in db.query(ProjectProgress).filter(
            ProjectProgress.subsidy_type_id == subsidy_type_id,
            ProjectProgress.village_name != "待分配",
        ).all():
            stages = json.loads(row.stages) if row.stages else []
            for s in stages:
                if s.get("name") == stage_name:
                    s["status"] = status
                    s["date"] = data.get("date", "")
            row.stages = json.dumps(stages, ensure_ascii=False)
        db.commit()
        return {"ok": True}

    if action == "add_stage_to_all":
        # 为所有记录添加新阶段
        stage = data.get("stage", {})
        for row in db.query(ProjectProgress).filter(
            ProjectProgress.subsidy_type_id == subsidy_type_id,
            ProjectProgress.village_name != "待分配",
        ).all():
            stages = json.loads(row.stages) if row.stages else []
            if not any(s.get("name") == stage.get("name") for s in stages):
                stages.append(stage)
            row.stages = json.dumps(stages, ensure_ascii=False)
        db.commit()
        return {"ok": True}

    if action == "sync_leaders":
        leads = {}
        for c in db.query(VillageContact).filter(VillageContact.is_agri_lead == True).all():
            leads[c.village_id] = c
        updated = 0
        for row in db.query(ProjectProgress).filter(
            ProjectProgress.subsidy_type_id == subsidy_type_id,
            ProjectProgress.village_name != "待分配",
        ).all():
            lead = leads.get(row.village_id)
            if lead:
                changed = False
                if lead.name and row.person_name != lead.name:
                    row.person_name = lead.name
                    changed = True
                if lead.phone and row.phone != lead.phone:
                    row.phone = lead.phone
                    changed = True
                if changed:
                    updated += 1
        db.commit()
        return {"ok": True, "updated": updated}

    if action == "swap_stages":
        name_a = data.get("stage_a", "")
        name_b = data.get("stage_b", "")
        if not name_a or not name_b:
            return {"error": "缺少 stage_a/stage_b"}, 400
        for row in db.query(ProjectProgress).filter(
            ProjectProgress.subsidy_type_id == subsidy_type_id,
            ProjectProgress.village_name != "待分配",
        ).all():
            stages = json.loads(row.stages) if row.stages else []
            i = next((idx for idx, s in enumerate(stages) if s.get("name") == name_a), -1)
            j = next((idx for idx, s in enumerate(stages) if s.get("name") == name_b), -1)
            if i >= 0 and j >= 0:
                stages[i], stages[j] = stages[j], stages[i]
            row.stages = json.dumps(stages, ensure_ascii=False)
        db.commit()
        return {"ok": True}

    return {"error": "unknown action"}, 400


@router.delete("/{subsidy_type_id}/{village_id}")
def delete_progress(
    subsidy_type_id: int,
    village_id: int,
    db: Session = Depends(get_db),
):
    record = db.query(ProjectProgress).filter(
        ProjectProgress.subsidy_type_id == subsidy_type_id,
        ProjectProgress.village_id == village_id,
    ).first()
    if record:
        db.delete(record)
        db.commit()
    return {"ok": True}


@router.post("/{subsidy_type_id}/batch-delete")
def batch_delete_progress(
    subsidy_type_id: int,
    data: dict,
    db: Session = Depends(get_db),
):
    """批量删除进度记录（传入 village_ids 列表）"""
    village_ids = data.get("village_ids", [])
    if not village_ids:
        return {"error": "缺少 village_ids"}, 400
    deleted = db.query(ProjectProgress).filter(
        ProjectProgress.subsidy_type_id == subsidy_type_id,
        ProjectProgress.village_id.in_(village_ids),
    ).delete(synchronize_session=False)
    db.commit()
    return {"ok": True, "deleted": deleted}


@router.post("/{subsidy_type_id}/delete-stage")
def delete_stage(
    subsidy_type_id: int,
    data: dict,
    db: Session = Depends(get_db),
):
    """删除指定阶段（所有记录）"""
    stage_name = data.get("stage_name", "")
    if not stage_name:
        return {"error": "缺少 stage_name"}, 400

    records = db.query(ProjectProgress).filter(
        ProjectProgress.subsidy_type_id == subsidy_type_id
    ).all()

    updated = 0
    for rec in records:
        stages = json.loads(rec.stages) if rec.stages else []
        new_stages = [s for s in stages if s.get("name") != stage_name]
        if len(new_stages) != len(stages):
            rec.stages = json.dumps(new_stages, ensure_ascii=False)
            updated += 1

    db.commit()
    return {"ok": True, "updated": updated, "message": f"已从 {updated} 个村删除阶段「{stage_name}」"}


@router.post("/{subsidy_type_id}/scan-files")
def scan_files_for_progress(
    subsidy_type_id: int,
    data: dict,
    db: Session = Depends(get_db),
):
    """
    扫描本地目录，按子文件夹匹配阶段，按文件名匹配村名。
    目录结构: D:/材料/宣传动员/红星村.pdf
    data: { path: "D:/材料", stage_name: "宣传动员" }
    """
    import os
    from datetime import date as date_type

    scan_path = data.get("path", "")
    stage_name = data.get("stage_name", "")

    if not scan_path or not os.path.isdir(scan_path):
        return {"error": f"目录不存在: {scan_path}"}, 400
    if not stage_name:
        return {"error": "缺少 stage_name"}, 400

    # 查找匹配阶段名的子文件夹；若路径本身就有文件则直接使用
    stage_dir = None
    matched_stage = ""
    try:
        for entry in os.listdir(scan_path):
            full = os.path.join(scan_path, entry)
            if os.path.isdir(full):
                if stage_name in entry or entry in stage_name:
                    stage_dir = full
                    matched_stage = entry
                    break
        # 没找到子文件夹时，检查路径本身是否就直接包含了文件
        if not stage_dir:
            files_in_root = [f for f in os.listdir(scan_path) if os.path.isfile(os.path.join(scan_path, f))]
            if files_in_root:
                stage_dir = scan_path
                matched_stage = os.path.basename(scan_path)
    except Exception as e:
        return {"error": f"读取目录失败: {e}"}, 500

    if not stage_dir:
        return {"error": f"未找到匹配阶段「{stage_name}」的子文件夹，路径下也无文件"}, 400

    # 获取阶段文件夹下的文件名（不含扩展名）
    file_names: set[str] = set()
    try:
        for f in os.listdir(stage_dir):
            name, _ = os.path.splitext(f)
            file_names.add(name.strip())
    except Exception as e:
        return {"error": f"读取阶段目录失败: {e}"}, 500

    # 获取该项目的所有进度记录
    records = db.query(ProjectProgress).filter(
        ProjectProgress.subsidy_type_id == subsidy_type_id
    ).all()

    updated = 0
    today_str = date_type.today().isoformat()

    for rec in records:
        stages = json.loads(rec.stages) if rec.stages else []
        village = rec.village_name

        # 检查村名是否出现在文件名中
        found = any(village in f for f in file_names)

        # 查找或创建对应阶段
        stage_updated = False
        for s in stages:
            if s.get("name") == stage_name:
                if found and s.get("status") != "done":
                    s["status"] = "done"
                    s["date"] = today_str
                    stage_updated = True
                elif not found and s.get("status") == "done":
                    s["status"] = "none"
                    s["date"] = ""
                    stage_updated = True
                break
        else:
            status = "done" if found else "none"
            stages.append({"name": stage_name, "status": status, "date": today_str if found else "", "note": ""})
            stage_updated = True

        if stage_updated:
            rec.stages = json.dumps(stages, ensure_ascii=False)
            updated += 1

    db.commit()

    return {
        "ok": True,
        "updated": updated,
        "total": len(records),
        "scanned_files": len(file_names),
        "matched_stage_dir": matched_stage,
        "message": f"扫描完成：阶段「{matched_stage}」下 {len(file_names)} 个文件，更新 {updated}/{len(records)} 个村",
    }


def _serialize(r: ProjectProgress) -> dict:
    return {
        "id": r.id,
        "subsidy_type_id": r.subsidy_type_id,
        "village_id": r.village_id,
        "village_name": r.village_name,
        "person_name": r.person_name or "",
        "phone": r.phone or "",
        "stages": json.loads(r.stages) if r.stages else [],
        "note": r.note or "",
        "updated_at": r.updated_at.isoformat() if r.updated_at else "",
    }

File: subsidy-electron/electron/ipc/project-progress.ts
import { ipcMain } from 'electron'
import { getDb } from '../database/connection'
import { success, errorResponse } from './response'
import { readdirSync, existsSync, statSync } from 'fs'
import { join } from 'path'

export function registerProjectProgressHandlers(): void {
  const db = () => getDb()

  // ensure table
  try {
    db().runRaw(`
      CREATE TABLE IF NOT EXISTS project_progress (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        subsidy_type_id INTEGER NOT NULL,
        village_id INTEGER NOT NULL,
        village_name TEXT DEFAULT '',
        person_name TEXT DEFAULT '',
        phone TEXT DEFAULT '',
        stages TEXT DEFAULT '[]',
        note TEXT DEFAULT '',
        updated_at TEXT DEFAULT (datetime('now','localtime')),
        UNIQUE(subsidy_type_id, village_id)
      )
    `)
  } catch { /* ignore */ }

  // ── 获取某个项目的所有村进度 ──
  ipcMain.handle('project-progress:get', (_e, projectId: number) => {
    try {
      const rows = db().allRaw<Record<string, unknown>>(`
        SELECT * FROM project_progress WHERE subsidy_type_id = ? ORDER BY village_name
      `, projectId)

      const records = rows.map(r => ({
        ...r,
        stages: safeParse(r.stages as string, []),
      }))
      return success(records)
    } catch (e) {
      return errorResponse(String(e))
    }
  })

  // ── 保存单村进度 ──
  ipcMain.handle('project-progress:save', (_e, payload: any) => {
    try {
      const { projectId, village_id, village_name, person_name, phone, stages, note } = payload
      const stagesJson = JSON.stringify(stages || [])

      const existing = db().getRaw<{ id: number }>(
        'SELECT id FROM project_progress WHERE subsidy_type_id = ? AND village_id = ?',
        projectId, village_id
      )

      if (existing) {
        db().runRaw(
          `UPDATE project_progress SET village_name=?, person_name=?, phone=?, stages=?, note=?, updated_at=datetime('now','localtime') WHERE id=?`,
          village_name || '', person_name || '', phone || '', stagesJson, note || '', existing.id
        )
      } else {
        db().runRaw(
          `INSERT INTO project_progress (subsidy_type_id, village_id, village_name, person_name, phone, stages, note) VALUES (?,?,?,?,?,?,?)`,
          projectId, village_id, village_name || '', person_name || '', phone || '', stagesJson, note || ''
        )
      }
      return success({ message: '已保存' })
    } catch (e) {
      return errorResponse(String(e))
    }
  })

  // ── 批量操作 ──
  ipcMain.handle('project-progress:batch', (_e, payload: any) => {
    try {
      const { projectId, action } = payload
      const rows = db().allRaw<Record<string, unknown>>(
        'SELECT * FROM project_progress WHERE subsidy_type_id = ?', projectId
      )

      switch (action) {
        case 'init': {
          // 从村庄表初始化所有村
          const villages = db().allRaw<{ id: number; village_name: string; group_no: number }>(
            'SELECT vg.*, v.village_name FROM village_group vg JOIN village v ON vg.village_id = v.id ORDER BY v.village_name, vg.group_no'
          )
          let added = 0
          for (const v of villages) {
            const exists = rows.find(r => r.village_id === v.id)
            if (!exists) {
              db().runRaw(
                `INSERT INTO project_progress (subsidy_type_id, village_id, village_name, stages) VALUES (?,?,?,?)`,
                projectId, v.id, `${v.village_name}${formatGroupNoSimple(v.group_no)}`, '[]'
              )
              added++
            }
          }
          return success({ message: `已初始化 ${added} 个村` })
        }

        case 'sync_leaders': {
          // 从村庄表同步负责人信息
          let updated = 0
          for (const r of rows) {
            const leader = db().getRaw<{ real_name: string; phone: string }>(`
              SELECT fp.real_name, fp.phone FROM farmer_profile fp
              JOIN family_household hh ON fp.household_id = hh.id
              WHERE hh.village_id = (SELECT village_id FROM village_group WHERE id = ?)
              AND fp.relation = '本人'
              LIMIT 1
            `, r.village_id)
            if (leader) {
              db().runRaw(
                `UPDATE project_progress SET person_name=?, phone=?, updated_at=datetime('now','localtime') WHERE id=?`,
                leader.real_name || '', leader.phone || '', r.id
              )
              updated++
            }
          }
          return success({ message: `已同步`, updated })
        }

        case 'add_stage_to_all': {
          const { stage } = payload
          for (const r of rows) {
            const stages = safeParse(r.stages as string, [])
            if (!stages.find((s: any) => s.name === stage.name)) {
              stages.push(stage)
              db().runRaw(
                `UPDATE project_progress SET stages=?, updated_at=datetime('now','localtime') WHERE id=?`,
                JSON.stringify(stages), r.id
              )
            }
          }
          return success({ message: `已添加阶段「${stage.name}」` })
        }

        case 'batch_stage': {
          const { stage_name, status, date } = payload
          for (const r of rows) {
            const stages = safeParse(r.stages as string, [])
            const idx = stages.findIndex((s: any) => s.name === stage_name)
            if (idx >= 0) {
              stages[idx] = { ...stages[idx], status, date: date || stages[idx].date }
              db().runRaw(
                `UPDATE project_progress SET stages=?, updated_at=datetime('now','localtime') WHERE id=?`,
                JSON.stringify(stages), r.id
              )
            }
          }
          return success({ message: `已批量更新「${stage_name}」` })
        }

        case 'swap_stages': {
          const { stage_a, stage_b } = payload
          for (const r of rows) {
            const stages = safeParse(r.stages as string, [])
            const ia = stages.findIndex((s: any) => s.name === stage_a)
            const ib = stages.findIndex((s: any) => s.name === stage_b)
            if (ia >= 0 && ib >= 0) {
              ;[stages[ia], stages[ib]] = [stages[ib], stages[ia]]
              db().runRaw(
                `UPDATE project_progress SET stages=?, updated_at=datetime('now','localtime') WHERE id=?`,
                JSON.stringify(stages), r.id
              )
            }
          }
          return success({ message: '已交换阶段顺序' })
        }

        default:
          return errorResponse('未知操作: ' + action)
      }
    } catch (e) {
      return errorResponse(String(e))
    }
  })

  // ── 删除阶段 ──
  ipcMain.handle('project-progress:deleteStage', (_e, payload: any) => {
    try {
      const { projectId, stage_name } = payload
      const rows = db().allRaw<Record<string, unknown>>(
        'SELECT * FROM project_progress WHERE subsidy_type_id = ?', projectId
      )
      for (const r of rows) {
        const stages = safeParse(r.stages as string, [])
        const filtered = stages.filter((s: any) => s.name !== stage_name)
        if (filtered.length !== stages.length) {
          db().runRaw(
            `UPDATE project_progress SET stages=?, updated_at=datetime('now','localtime') WHERE id=?`,
            JSON.stringify(filtered), r.id
          )
        }
      }
      return success({ message: `已删除阶段「${stage_name}」` })
    } catch (e) {
      return errorResponse(String(e))
    }
  })

  // ── 文件扫描 ──
  ipcMain.handle('project-progress:scanFiles', (_e, payload: any) => {
    try {
      const { projectId, path: scanPath, stage_name } = payload
      if (!existsSync(scanPath)) return errorResponse('目录不存在: ' + scanPath)

      const files = readdirSync(scanPath).filter(f => {
        try { return statSync(join(scanPath, f)).isFile() } catch { return false }
      })

      const rows = db().allRaw<Record<string, unknown>>(
        'SELECT * FROM project_progress WHERE subsidy_type_id = ?', projectId
      )

      let matched = 0
      for (const r of rows) {
        const vname = String(r.village_name || '')
        const found = files.some(f => {
          const basename = f.replace(/\.[^.]+$/, '') // strip extension
          return basename.includes(vname) || vname.includes(basename)
        })
        if (found) {
          const stages = safeParse(r.stages as string, [])
          const idx = stages.findIndex((s: any) => s.name === stage_name)
          if (idx >= 0) {
            stages[idx] = { ...stages[idx], status: 'done', date: new Date().toISOString() }
            db().runRaw(
              `UPDATE project_progress SET stages=?, updated_at=datetime('now','localtime') WHERE id=?`,
              JSON.stringify(stages), r.id
            )
            matched++
          }
        }
      }

      return success({ message: `扫描完成：${files.length} 个文件，匹配 ${matched} 个村` })
    } catch (e) {
      return errorResponse(String(e))
    }
  })
}

function safeParse(val: string, fallback: unknown): any {
  try { return JSON.parse(val || '[]') } catch { return fallback }
}

function formatGroupNoSimple(n: number): string {
  const map = ['零', '一', '二', '三', '四', '五', '六', '七', '八', '九', '十']
  if (n >= 1 && n <= 10) return map[n] + '组'
  return n + '组'
}
